src/osmet/mod.rs

```diff
@@ -0,0 +1,579 @@
+// Copyright 2020 Red Hat, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::hash_map::Entry;
+use std::collections::HashMap;
+use std::convert::TryInto;
+use std::fs::{File, OpenOptions};
+use std::io::{self, Seek, SeekFrom, Write};
+use std::os::unix::fs::FileTypeExt;
+use std::path::{Path, PathBuf};
+
+use error_chain::bail;
+use nix::mount;
+use serde::{Deserialize, Serialize};
+use walkdir::WalkDir;
+use xz2::write::XzEncoder;
+
+use crate::blockdev::*;
+use crate::cmdline::*;
+use crate::errors::*;
+
+mod fiemap;
+mod file;
+mod io_helpers;
+mod unpacker;
+
+use crate::osmet::fiemap::*;
+use crate::osmet::io_helpers::*;
+use crate::osmet::file::*;
+use crate::osmet::unpacker::*;
+
+// just re-export OsmetUnpacker
+pub use crate::osmet::unpacker::OsmetUnpacker;
+
+#[derive(Serialize, Deserialize, Debug)]
+struct Mapping {
+    extent: Extent,
+    object: Sha256Digest,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+struct OsmetPartition {
+    start_offset: u64,
+    end_offset: u64,
+    mappings: Vec<Mapping>,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+struct Osmet {
+    /// Partitions for which we've registered mappings.
+    partitions: Vec<OsmetPartition>,
+    /// Checksum of the final disk image. Used for unpacking verification.
+    checksum: Sha256Digest,
+    /// Size of the final disk image. Used for unpacking verification.
+    size: u64,
+    /// Required sector size of target block device during unpacking.
+    sector_size: u32,
```

Comment:
We also have a dependency on the CPU architecture of the target, so we should record that as well.  It's possible for `coreos-installer` to write an image with a different arch than its own, e.g. when writing to a USB stick.

Reply:
I think it's worth clarifying that technically neither packing nor unpacking have any dependency on the intended CPU architecture of the source disk image. I.e. it's entirely feasible to run `coreos-installer osmet pack` on aarch64 while pointing it at a mounted x86_64 disk.

That said, I agree with you we should pick up that info as a sanity-check on unpacking. The shortcut I took for now is to just assume that the node which runs `coreos-installer osmet pack` is on the same architecture as the one intended for the disk we're packing. This is true today because cosa doesn't support cross-building.

Anyway, with that, `coreos-installer install --architecture <some other arch>` won't use osmet since it doesn't match.

I also added a comment about using `coreos-assembler.basearch` from the base commit metadata would be more correct (or we could just add `basearch` to the aleph JSON file too).